src/hooks/useOrders.ts

```diff
@@ -27,19 +34,165 @@ function filterDeletedOrders(orders: AuctionElement[]): AuctionElement[] {
+    ): Promise<void> => {
+      if (userAddress === undefined || networkId === undefined || offset === undefined) {
+        return
+      }
+
+      await exchangeApi.getOrdersPaginated({ userAddress, networkId, offset }).then(({ orders, nextIndex }) => {
+        if (orders.length) {
+          // Apply filters (remove deleted orders)
+          const filteredOrders = filterDeletedOrders(orders)
+
+          // Store new orders, if any
+          if (offset === 0) {
+            // fresh start/refresh: replace whatever is stored
+            setOrdersState(curr => updateOrdersState(networkId, userAddress, curr, filteredOrders, false))
+          } else if (filteredOrders.length) {
+            // incremental update: append
+            setOrdersState(curr => updateOrdersState(networkId, userAddress, curr, filteredOrders, true))
+          }
+        } else if (offset === 0 && areThereOrders) {
+          // There were orders. We fetched again from the beginning, and now there are none (in the first page). Set to []
+          // Example: switching network or address.
+          setOrdersState(curr => updateOrdersState(networkId, userAddress, curr, [], false))
+        }
+
+        // `nextIndex` can be `undefined`, which means there are no more pages
+        if (nextIndex) {
+          // There are more, keep going
+          _fetchOrders(userAddress, networkId, nextIndex)
+        } else {
+          // We are done, remember where we stopped for next time
+          setOffset(offset + orders.length)
+        }
+      })
+    },
+    [],
+  )
+
+  const fetchOrdersAndHandleErrors = useCallback(
+    async (
+      userAddress: string | undefined,
+      networkId: number | undefined,
+      offset: number | undefined,
+      areThereOrders?: boolean,
+    ): Promise<void> => {
+      setIsLoading(true)
+      try {
+        await _fetchOrders(userAddress, networkId, offset, areThereOrders)
+      } catch (e) {
+        console.log('Failed to fetch orders', e)
+        // TODO: inform user
+      } finally {
+        setIsLoading(false)
+      }
+    },
+    [],
+  )
+
+  useEffect(() => {
+    setLastBlockNumber(blockNumber)
+  }, [blockNumber, setLastBlockNumber])
+
+  useEffect(() => {
+    setLastNetworkId(networkId)
+  }, [networkId, setLastNetworkId])
+
+  useEffect(() => {
+    setLastUserAddress(userAddress)
+  }, [userAddress, setLastUserAddress])
 
   useEffect(() => {
-    userAddress &&
-      networkId &&
-      exchangeApi
-        .getOrders({ userAddress, networkId })
-        .then(filterDeletedOrders)
-        .then(setOrders)
-    // updating list of orders on every block by listening on `blockNumber`
-  }, [networkId, setOrders, userAddress, blockNumber])
-
-  return orders
+    // whenever there's a state change (new block, network /address change)
+
+    if (!isLoading && networkId && userAddress) {
+      // should try to update orders
+      if (networkId !== lastNetworkId || userAddress !== lastUserAddress) {
```

Comment:
I don't quite get the reason for keeping track of `lastNetworkId, lastUserAddress`
How about separate useEffects
```
[networkId, userAddress] => setOffset(0)
[blockNumber] => fetch from offset
```

Reply:
This is complicated.

I was using two useEffect. 1 for blockNumber one for network/address.
This caused on the initial load to trigger two recursions in parallel, because useEffect is always run on the first render (right?), which made the list of orders contain duplicates.

And this is sort of why tracking I'm tracking lastNetworkId, lastUserAddress.

Unless you know how I could have 2 useEffect in a way that one doesn't clash with the other?